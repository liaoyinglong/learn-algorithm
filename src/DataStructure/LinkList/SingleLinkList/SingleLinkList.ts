/**
 * 单列表
 */
export class SingleLinkList<T> {
  private len = 0;
  private head: Node<T> | null = null;

  add(item: T) {
    const node = new Node(item);
    if (!this.head) {
      this.head = node;
      this.len = 1;
      return;
    }

    let last = this.head;

    while (last.next) {
      last = last.next;
    }
    last.next = node;
    this.len++;
  }

  get(targetIndex: number) {
    if (targetIndex < 0 || targetIndex > this.len) {
      return;
    }
    let i = 0;
    let node = this.head!;
    do {
      if (i === targetIndex) {
        return node;
      }
      i++;
      node = node.next!;
    } while (node);
  }

  indexOf(item: T) {
    let node = this.head;
    let i = 0;

    while (node) {
      if (node.data === item) {
        return i;
      }
      node = node.next;
      i++;
    }
    return -1;
  }
}

class Node<T> {
  next: null | Node<T> = null;

  constructor(public data: T) {}
}
