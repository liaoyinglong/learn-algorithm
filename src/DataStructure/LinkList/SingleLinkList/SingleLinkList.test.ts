import { SingleLinkList } from "./SingleLinkList";

describe("SingleLinkList", () => {
  let list: SingleLinkList<number | string>;

  beforeEach(() => {
    list = new SingleLinkList<number | string>();
    list.add("a");
    list.add("b");
    list.add("c");
  });

  function expectLen(len: number) {
    expect(list.length).toBe(len);
  }

  it("add should be correctly", () => {
    expect(list).toMatchSnapshot();
    expectLen(3);
  });

  it("get should be correctly ", () => {
    expect(list.get(0)).toMatchSnapshot();
    expect(list.get(2)).toMatchSnapshot();
  });

  it("indexOf should be correctly ", () => {
    expect(list.indexOf("a")).toBe(0);
    expect(list.indexOf("b")).toBe(1);
    expect(list.indexOf("c")).toBe(2);
    expect(list.indexOf("d")).toBe(-1);
  });
  it("has should be correctly ", () => {
    expect(list.has("a")).toBe(true);
    expect(list.has("b")).toBe(true);
    expect(list.has("c")).toBe(true);
    expect(list.has("d")).toBe(false);
  });
  it("inset should be correctly ", () => {
    list.inset("d", 0);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(4);

    list.inset("f", 1);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(5);

    list.inset("g", 1);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(6);
  });
  it("inset should return false", () => {
    expect(list.inset("g", 1111)).toBe(false);
    expectLen(3);
  });
  it("remove should be correctly ", () => {
    list.remove(1);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(2);

    list.remove(1);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(1);

    list.remove(0);
    expect(list.get(0)).toMatchSnapshot();
    expectLen(0);
  });
});
