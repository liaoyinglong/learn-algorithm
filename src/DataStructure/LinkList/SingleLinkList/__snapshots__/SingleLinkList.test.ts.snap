// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleLinkList add should be correctly 1`] = `
SingleLinkList {
  "head": Node {
    "data": "a",
    "next": Node {
      "data": "b",
      "next": Node {
        "data": "c",
        "next": null,
      },
    },
  },
  "len": 3,
}
`;

exports[`SingleLinkList get should be correctly  1`] = `
Node {
  "data": "a",
  "next": Node {
    "data": "b",
    "next": Node {
      "data": "c",
      "next": null,
    },
  },
}
`;

exports[`SingleLinkList get should be correctly  2`] = `
Node {
  "data": "c",
  "next": null,
}
`;
